File: src/api/routes/tasks.ts
import { Router, Response } from 'express';
import { randomUUID } from 'crypto';
import { authenticate, AuthRequest } from '../middleware/auth';
import {
  createTask,
  getTaskById,
  listTasks,
  getTaskLogs,
} from '../../db/taskRepository';
import { enqueue } from '../../queue/queue';
import { CreateTaskRequest, TaskPriority, TaskStatus, TaskType } from '../../types';

const router = Router();

const VALID_TYPES: TaskType[] = ['email', 'image_resize', 'report_generation', 'data_export', 'notification'];
const VALID_PRIORITIES: TaskPriority[] = ['high', 'normal', 'low'];

// POST /api/tasks — enqueue a new task
router.post('/', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const body = req.body as CreateTaskRequest;

  if (!body.type || !VALID_TYPES.includes(body.type)) {
    res.status(400).json({
      error: `Invalid task type. Must be one of: ${VALID_TYPES.join(', ')}`,
    });
    return;
  }
  if (!body.payload || typeof body.payload !== 'object') {
    res.status(400).json({ error: 'payload must be a JSON object' });
    return;
  }
  if (body.priority && !VALID_PRIORITIES.includes(body.priority)) {
    res.status(400).json({
      error: `Invalid priority. Must be one of: ${VALID_PRIORITIES.join(', ')}`,
    });
    return;
  }

  const id = randomUUID();
  const priority = body.priority ?? 'normal';
  const maxRetries = body.max_retries ?? 3;

  try {
    const task = await createTask({
      id,
      type: body.type,
      priority,
      payload: body.payload,
      max_retries: maxRetries,
      webhook_url: body.webhook_url,
    });

    // Push to the appropriate Redis queue
    await enqueue(id, body.type, priority);

    res.status(202).json({
      task_id: task.id,
      status: task.status,
      priority: task.priority,
      type: task.type,
      created_at: task.created_at,
      message: 'Task enqueued successfully',
    });
  } catch {
    res.status(500).json({ error: 'Failed to enqueue task' });
  }
});

// GET /api/tasks — list tasks with optional filters
router.get('/', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const status = req.query.status as TaskStatus | undefined;
  const type = req.query.type as string | undefined;
  const limit = parseInt(req.query.limit as string) || 20;
  const offset = parseInt(req.query.offset as string) || 0;

  if (limit > 100) {
    res.status(400).json({ error: 'limit cannot exceed 100' });
    return;
  }

  try {
    const { tasks, total } = await listTasks({ status, type, limit, offset });
    res.json({
      tasks,
      pagination: { total, limit, offset, has_more: offset + limit < total },
    });
  } catch {
    res.status(500).json({ error: 'Failed to fetch tasks' });
  }
});

// GET /api/tasks/:id — get single task with logs
router.get('/:id', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  try {
    const task = await getTaskById(req.params['id'] as string);
    if (!task) {
      res.status(404).json({ error: 'Task not found' });
      return;
    }
    const logs = await getTaskLogs(task.id);
    res.json({ ...task, logs });
  } catch {
    res.status(500).json({ error: 'Failed to fetch task' });
  }
});

export default router;

File: app/api/routes/tasks.py
import uuid
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession
from app.db.base import get_db
from app.db.repository import create_task, get_task, list_tasks, get_task_logs
from app.queue.queue import enqueue
from app.schemas.task import (
    TaskCreate, TaskResponse, TaskDetailResponse,
    TaskListResponse, TaskLogResponse, TaskStatus,
)
from app.api.deps import get_current_user

router = APIRouter(prefix="/tasks", tags=["tasks"])


@router.post("", response_model=TaskResponse, status_code=status.HTTP_202_ACCEPTED)
async def create_task_endpoint(
    body: TaskCreate,
    db: AsyncSession = Depends(get_db),
    _user: dict = Depends(get_current_user),
):
    task_id = str(uuid.uuid4())
    task = await create_task(
        db,
        id=task_id,
        type=body.type.value,
        priority=body.priority.value,
        payload=body.payload,
        max_retries=body.max_retries,
        webhook_url=body.webhook_url,
    )
    await enqueue(task_id, body.type.value, body.priority.value)
    return task


@router.get("", response_model=TaskListResponse)
async def list_tasks_endpoint(
    status: Optional[str] = Query(None),
    type: Optional[str] = Query(None),
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db),
    _user: dict = Depends(get_current_user),
):
    tasks, total = await list_tasks(db, status=status, task_type=type, limit=limit, offset=offset)
    return TaskListResponse(
        tasks=tasks, total=total, limit=limit, offset=offset,
        has_more=offset + limit < total,
    )


@router.get("/{task_id}", response_model=TaskDetailResponse)
async def get_task_endpoint(
    task_id: str,
    db: AsyncSession = Depends(get_db),
    _user: dict = Depends(get_current_user),
):
    task = await get_task(db, task_id)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    logs = await get_task_logs(db, task_id)
    log_responses = [
        TaskLogResponse(
            id=log.id, task_id=log.task_id, worker_id=log.worker_id,
            event=log.event, message=log.message,
            metadata_=log.metadata_, created_at=log.created_at,
        )
        for log in logs
    ]
    return TaskDetailResponse(
        **{c.key: getattr(task, c.key) for c in task.__table__.columns},
        logs=log_responses,
    )
